spec-0007/index.md

```diff
@@ -0,0 +1,82 @@
+---
+title: "SPEC 7 — Seeding pseudo-random number generation"
+date: 2023-04-19
+author:
+  - Other participants in the discussion <not.yet@named.org>"
+  - "Stéfan van der Walt <stefanv@berkeley.edu>"
+discussion: https://github.com/scipy/scipy/issues/14322
+endorsed-by:
+---
+
+## Description
+
+<!--
+Briefly and clearly describe the proposal.
+Explain the general need and the advantages of this specific proposal.
+If relevant, include examples of how the new functionality would be used,
+intended use-cases, and pseudo-code illustrating its use.
+-->
+
+There is disparity in the APIs libraries use to seed random number generation.
+The goal of this SPEC is suggest a single, pragmatic API for the ecosystem, taking into account technical and historical factors.
+
+### Concepts
+
+- `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`np.random.default_rng`) uses PCG64.
+- `Generator`: Derives pseudo-random numbers from the bits produced by a `BitGenerator`.
+- `RandomState`: a [legacy object in NumPy](https://numpy.org/doc/stable/reference/random/index.html), similar to `Generator`, that produces random numbers based on the Mersenne Twister.
+
+### Constraints
+
+NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
+Common keyword arguments include `random_state` and `seed`.
+In practice, the seed is unfortunately also often controlled using `np.random.seed`.
+
+## Implementation
+
+<!--
+Discuss how this would be implemented.
+-->
+
+The new API takes into account legacy behavior in packages such as scikit-learn (see `sklearn.utils.check_random_state`), which works as follows:
+
+1. Because `np.random.seed` is so often used in practice, no seed means
+   using the global `RandomState` object, `np.random.mtrand._rand`.
+2. (Option a) When a seed is provided, a `RandomState` object is initialized with that seed.
+3. (Option b) When a seed is provided, a `Generator` object is initialized with that seed.
+4. If an instance of `RandomState` is provided, it is used as-is.
+5. If an instance of `Generator` is provided, it is used as-is.
```

Comment:
This is describing the current state in some libraries. But is it where we want to see this 10 years from now?

I am personally against any global state and advertising of any "legacy" behaviours.

Reply:
I also feel we may want to think about a new keyword argument instead, that adopts recommended best practices instead.

Reply:
I am not sure I see how that could help.

If you add a `rng` to a function which has `seed` or `random_state` we don't avoid raising some warning about deprecation, etc.

Reply:
The big problem with `random_state` is that it allows for `None`, which then grabs global state. So, that will always conflict with an `rng=None` kwarg.

Reply:
To me, the ideal API (using todays tool at least) would be that `random_state=None` would give you `np.random.default_rng()`.

There is also the crazy thought, which I kind of like, from @ilayn: do not accept integers, only a `Generator` (or other object). Point being, you must provide a RNG if you want any reproducibility.

Reply:
I don't think we can do that, though, because it would be a backward incompatible change?

Reply:
Both integer and global state behaviour are BC issues yep.

It would be interesting to see how in real life `Generator` was painful to move to, taking into account a large sample of projects, folks, etc.

Reply:
Yes, and don't expect the same favors from commercial code. It's shackles that we choose to put on, sometimes at great time cost to  (oftentimes volunteer) developers.